Lib/extractor/formats/opentype.py

```diff
@@ -1078,3 +1012,185 @@ def extractOpenTypeFeatures(source):
+
+    allAnchors = []
+    seenLookups = set()
+    for script in scriptOrder:
+        for lookupIndex in lookupIndexes[script]:
+            if lookupIndex in seenLookups:
+                continue
+            seenLookups.add(lookupIndex)
+            anchorsForThisLookup = _gatherAnchorsForLookup(gpos, lookupIndex)
+            allAnchors = allAnchors + anchorsForThisLookup
+    return allAnchors
+
+
+def _gatherAnchorsForLookup(gpos, lookupIndex):
+    """
+    Gather the anchor data for a particular lookup.
+    Returns a list of anchor group data dicts in the following format:
+        {
+            "baseAnchors": {"A": {"x": 672, "y": 1600}, "B": {"x": 624, "y": 1600}},
+            "markAnchors": {'gravecomb': {'x': -400, 'y': 1500}, 'acutecomb': {'x': -630, 'y': 1500}},
+        }
+    """
+    allAnchorGroups = []
+    lookup = gpos.LookupList.Lookup[lookupIndex]
+    # Type 4 are mark-to-base attachment lookups, type 6 are mark-to-mark ones, type 9 are extended lookups.
+    if lookup.LookupType not in (4, 6, 9):
+        return allAnchorGroups
+    if lookup.LookupType == 9 and lookup.SubTable[0].ExtensionLookupType not in (4,6):
+        return allAnchorGroups
+    for subtableIndex, subtable in enumerate(lookup.SubTable):
+        if (subtable.Format != 1):
+            print(f"  Skipping Anchor lookup subtable of unknown format {subtable.Format}.")
+            continue
+        if (lookup.LookupType == 9):
+            subtable = subtable.ExtSubTable
+        subtableAnchors = _handleAnchorLookupType4Format1(subtable)
+        allAnchorGroups.append(subtableAnchors)
+    return allAnchorGroups
+
+
+def _handleAnchorLookupType4Format1(subtable):
+    """
+    Extract anchors from a Lookup Type 4 Format 1.
+    """
+    anchors = {
+        "baseAnchors": {},
+        "markAnchors": {},
+    }
+
+    if subtable.LookupType not in (4, 6):
+        print(f"  Skipping Anchor lookup subtable with unsupported LookupType {subtable.LookupType}.")
+        return anchors
+
+    subtableIsType4 = subtable.LookupType == 4
+
+    baseCoverage = subtable.BaseCoverage.glyphs if subtableIsType4 else subtable.Mark2Coverage.glyphs
+    markCoverage = subtable.MarkCoverage.glyphs if subtableIsType4 else subtable.Mark1Coverage.glyphs
+
+    for baseRecordIndex, baseRecord in enumerate(subtable.BaseArray.BaseRecord if subtableIsType4 else subtable.Mark2Array.Mark2Record):
+        baseAnchor = baseRecord.BaseAnchor[0] if subtableIsType4 else baseRecord.Mark2Anchor[0]
+        anchors["baseAnchors"].update({baseCoverage[baseRecordIndex]: {"x": baseAnchor.XCoordinate, "y": baseAnchor.YCoordinate}})
+
+    for markRecordIndex, markRecord in enumerate(subtable.MarkArray.MarkRecord if subtableIsType4 else subtable.Mark1Array.MarkRecord):
+        markAnchor = markRecord.MarkAnchor
+        anchors["markAnchors"].update({markCoverage[markRecordIndex]: {"x": markAnchor.XCoordinate, "y": markAnchor.YCoordinate}})
+
+    return anchors
+
+
+# -------
```

Comment:
Moving these functions down makes it difficult to compare the previous to your changes with the diff, possible to move back up?

Reply:
Okay, done.

I would expect these shared methods to be moved to their own group in the end though.

Reply:
I'll have to look at the tests as well, but I'm too busy making use of this change on my PC right now. :smile: